tests/integration/upgrades/helpers.py

```diff
@@ -65,63 +96,348 @@ async def refresh(
+        ops_test,
+        "PUT",
+        f"https://{unit_ip}:9200/_cluster/settings",
+        payload={"persistent": {"cluster.routing.allocation.enable": "all"}},
+    )
+
+    time.sleep(5)
+
+    # get health
+    cluster_health_resp = await cluster_health(ops_test, unit_ip)
+    logger.info(f"Cluster health response: {cluster_health_resp["status"]}")
+    if cluster_health_resp["status"] == "red":
+        # identify problematic index
+        shards = await http_request(
+            ops_test,
+            "GET",
+            f"https://{unit_ip}:9200/_cat/shards?format=json&h=index,shard,state,unassigned.reason",
+        )
+
+        indices = set()
+        for shard in shards:
+            if (
+                shard.get("state") == "UNASSIGNED"
+                and shard.get("unassigned.reason") == "NODE_LEFT"
+            ):
+                indices.add(shard.get("index"))
+
+        # delete the indices
+        logger.info(f"Unassigned indices: {indices}")
+        for index in indices:
+            await http_request(
+                ops_test,
+                "DELETE",
+                f"https://{unit_ip}:9200/{index}",
+            )
+
+        cluster_health_resp = await cluster_health(ops_test, unit_ip)
+        logger.info(
+            f"Cluster health response after removing indices: {cluster_health_resp["status"]}"
         )
+    # add unit
+    logger.info("Adding new unit")
+    await ops_test.model.applications[app].add_unit(count=1)
+
+    # destroy rolled back unit
+    logger.info(f"Destroying unit `{app}/{rolled_back_unit_id}`")
+    await ops_test.model.destroy_unit(
+        f"{app}/{rolled_back_unit_id}", destroy_storage=True, force=True
+    )
+    await ops_test.model.block_until(
+        lambda: len(ops_test.model.applications[app].units) == len(units), timeout=300
+    )
+
+    remaining_units = await get_application_units(ops_test, app)
+    new_unit_id = sorted([unit.id for unit in remaining_units])[-1]
+    logger.info(f"Waiting for new unit {app}/{new_unit_id}...")
+    await wait_until_condition_on_units(
+        ops_test,
+        app=app,
+        condition=lambda units: any(
+            unit.workload_status.message
+            == "Requesting lock on operation: start"  # unit may be stuck waiting for lock
+            or unit.agent_status.value == "idle"
+            for unit in units
+            if unit.id == new_unit_id
+        ),
+        timeout=TIMEOUT,
+    )
+
+    # check if lock with departed unit
+    logger.info(f"Rolled back OpenSearch node: {rolled_back_node}")
+    lock_doc = await http_request(
+        ops_test,
+        "GET",
+        f"https://{unit_ip}:9200/.charm_node_lock/_doc/0",
+    )
+    if node_with_lock := lock_doc.get("_source", {}).get("unit-name"):
+        logger.info(f"Unit with lock: {node_with_lock}")
+
+        if node_with_lock == rolled_back_node:
+            logger.info("Deleting lock document")
+            lock_doc = await http_request(
```

Comment:
No need to assign lock_doc variable, we do not use it.
```suggestion
           await http_request(
```